0.9.12: 32bit arches: test_ndarray_memory_error raises ValueError instead of expected MemoryError

This is a Fedora package build from https://github.com/serge-sans-paille/pythran/pull/1831

On 32bit arches (armv7hl and i686), test_ndarray_memory_error raises `ValueError` instead of expected `MemoryError`.

```
=================================== FAILURES ===================================
____________________ TestNdarray.test_ndarray_memory_error _____________________
[gw4] linux -- Python 3.10.0 /usr/bin/python3
self = <pythran.tests.test_ndarray.TestNdarray testMethod=test_ndarray_memory_error>
    def test_ndarray_memory_error(self):
        with self.assertRaises(MemoryError):
>           self.run_test('def ndarray_memory_error(n): import numpy as np; return np.ones(n)',
                          999999999999,
                          ndarray_memory_error=[int])
pythran/tests/test_ndarray.py:22: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pythran/tests/__init__.py:313: in run_test
    python_ref = self.run_python(code, (name, copy.deepcopy(params)),
pythran/tests/__init__.py:173: in run_python
    ret_val = env[runas[0]](*runas[1])
<string>:1: in ndarray_memory_error
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
    @set_array_function_like_doc
    @set_module('numpy')
    def ones(shape, dtype=None, order='C', *, like=None):
        """
        Return a new array of given shape and type, filled with ones.
    
        Parameters
        ----------
        shape : int or sequence of ints
            Shape of the new array, e.g., ``(2, 3)`` or ``2``.
        dtype : data-type, optional
            The desired data-type for the array, e.g., `numpy.int8`.  Default is
            `numpy.float64`.
        order : {'C', 'F'}, optional, default: C
            Whether to store multi-dimensional data in row-major
            (C-style) or column-major (Fortran-style) order in
            memory.
        ${ARRAY_FUNCTION_LIKE}
    
            .. versionadded:: 1.20.0
    
        Returns
        -------
        out : ndarray
            Array of ones with the given shape, dtype, and order.
    
        See Also
        --------
        ones_like : Return an array of ones with shape and type of input.
        empty : Return a new uninitialized array.
        zeros : Return a new array setting values to zero.
        full : Return a new array of given shape filled with value.
    
    
        Examples
        --------
        >>> np.ones(5)
        array([1., 1., 1., 1., 1.])
    
        >>> np.ones((5,), dtype=int)
        array([1, 1, 1, 1, 1])
    
        >>> np.ones((2, 1))
        array([[1.],
               [1.]])
    
        >>> s = (2,2)
        >>> np.ones(s)
        array([[1.,  1.],
               [1.,  1.]])
    
        """
        if like is not None:
            return _ones_with_like(shape, dtype=dtype, order=order, like=like)
    
>       a = empty(shape, dtype, order)
E       ValueError: Maximum allowed dimension exceeded
/usr/lib/python3.10/site-packages/numpy/core/numeric.py:203: ValueError
```

--- a/pythran/tests/__init__.py
+++ b/pythran/tests/__init__.py
@@ -318,10 +318,12 @@ class TestEnv(unittest.TestCase):
         pythran_res = self.run_pythran(*run_pythran_args)
 
         if check_exception:
-            if pythran_res != python_ref:
+            if not issubclass(python_ref, pythran_res):
                 raise AssertionError(
                     "expected exception was %s, but received %s" %
                     (python_ref, pythran_res))
+            else:
+                return
 
         print("Python result: ", python_ref)
         print("Pythran result: ", pythran_res)

--- a/pythran/tests/test_ndarray.py
+++ b/pythran/tests/test_ndarray.py
@@ -13,15 +13,27 @@ except AttributeError:
     has_float128 = False
 
 
+huge = numpy.iinfo(numpy.intp).max // numpy.intp().itemsize
+
+def raisesMemoryError():
+    try:
+        numpy.ones(huge)
+        return False
+    except MemoryError:
+        return True
+    except ValueError:
+        return False
+
 
 @TestEnv.module
 class TestNdarray(TestEnv):
 
+    @unittest.skipIf(not raisesMemoryError(), "memory error not triggered on that arch")
     def test_ndarray_memory_error(self):
-        with self.assertRaises(MemoryError):
-            self.run_test('def ndarray_memory_error(n): import numpy as np; return np.ones(n)',
-                          999999999999,
-                          ndarray_memory_error=[int])
+        code = 'def ndarray_memory_error(n): import numpy as np; return np.ones(n)'
+        self.run_test(code, huge,
+                      ndarray_memory_error=[numpy.intp],
+                      check_exception=True)
 
     def test_ndarray_intc(self):
         self.run_test('def ndarray_intc(a): import numpy as np; return np.intc(a), np.array([a, a], dtype=np.intc)',

